app/src/main/java/agency/five/codebase/android/movieapp/data/repository/MovieRepositoryImpl.kt

```diff
@@ -0,0 +1,126 @@
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private val moviesByCategory: Map<MovieCategory, Flow<List<Movie>>> = MovieCategory.values()
+        .associateWith { movieCategory ->
+            flow {
+                val movieResponse = when (movieCategory) {
+                    MovieCategory.NOWPLAYING_MOVIES -> movieService.fetchNowPlayingMovies()
+                    MovieCategory.NOWPLAYING_TV -> movieService.fetchTopRatedMovies()
+                    MovieCategory.POPULAR_STREAMING -> movieService.fetchPopularMovies()
+                    MovieCategory.POPULAR_ONTV -> movieService.fetchTopRatedMovies()
+                    MovieCategory.POPULAR_FORRENT -> movieService.fetchNowPlayingMovies()
+                    MovieCategory.POPULAR_INTHEATHERS -> movieService.fetchUpcomingMovies()
+                    MovieCategory.UPCOMING_TODAY -> movieService.fetchUpcomingMovies()
+                    MovieCategory.UPCOMING_THISWEEK -> movieService.fetchPopularMovies()
+                }
+                emit(movieResponse.movies)
+            }.flatMapLatest { apiMovies ->
+                movieDao.favorites()
+                    .map { favoriteMovies ->
+                        apiMovies.map { apiMovie ->
+                            apiMovie.toMovie (isFavorite = favoriteMovies.any { it.id == apiMovie.id })
+                        }
+                    }
+            }.shareIn(
+                scope = CoroutineScope(bgDispatcher),
+                started = SharingStarted.WhileSubscribed(1000L),
+                replay = 1,
+            )
+        }
+
+    private val favorites = movieDao.favorites().map {
+        it.map { dbFavoriteMovie ->
+            Movie(
+                id = dbFavoriteMovie.id,
+                imageUrl = dbFavoriteMovie.posterUrl,
+                title = "",
+                overview = "",
+                isFavorite = true,
+            )
+        }
+    }.shareIn(
+        scope = CoroutineScope(bgDispatcher),
+        started = SharingStarted.WhileSubscribed(1000L),
+        replay = 1,
+    )
+
+    override fun movies(movieCategory: MovieCategory): Flow<List<Movie>> = moviesByCategory[movieCategory]!!
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    override fun movieDetails(movieId: Int): Flow<MovieDetails> = flow {
+        emit(movieService.fetchMovieDetails(movieId) to movieService.fetchMovieCredits(movieId))
+    }.flatMapLatest { (apiMovieDetails, apiMovieCredits) ->
+        movieDao.favorites()
+            .map { favoriteMovies ->
+                apiMovieDetails.toMovieDetails(
+                    isFavorite = favoriteMovies.any { it.id == apiMovieDetails.id },
+                    crew = apiMovieCredits.crew,
+                    cast = apiMovieCredits.cast
+                )
+            }
+    }.flowOn(bgDispatcher)
+
+    override fun favoriteMovies(): Flow<List<Movie>> = favorites
+
+    override suspend fun addMovieToFavorites(movieId: Int) {
+        runBlocking(bgDispatcher) {
```

Comment:
Remove `runBlocking`

Reply:
I need to run this on a background thread because accessing the database on main thread crashes the application.

Or is there maybe another way to do this?

Reply:
This is unusual because Room should internally change threads, for now just use `Dispatcher.IO` on `viewModelScope` where you call the repository functions

Reply:
Did this, app works.